lib/util/xpcapng.c

```diff
@@ -0,0 +1,630 @@
+		(packetid ? 12 : 0) + (queue ? 8 : 0) + (xdp_verdict ? 16 : 0);
+	rc = writev(pd->pd_fd, iov, i);
+	if (rc != epb_length)
+		return false;
+
+	return true;
+}
+
+/*****************************************************************************
+ * xpcapng_dump_open()
+ *****************************************************************************/
+struct xpcapng_dumper *xpcapng_dump_open(const char *file,
+					 const char *comment,
+					 const char *hardware,
+					 const char *os,
+					 const char *user_application)
+{
+	struct xpcapng_dumper *pd = NULL;
+
+	if (file == NULL) {
+		errno = EINVAL;
+		goto error_exit;
+	}
+
+	pd = calloc(sizeof(*pd), 1);
+	if (pd == NULL) {
+		errno = ENOMEM;
+		goto error_exit;
+	}
+	pd->pd_fd = -1;
+
+	if (strcmp(file, "-") == 0) {
+		pd->pd_fd = STDOUT_FILENO;
+	} else {
+		pd->pd_fd = open(file, O_WRONLY | O_CREAT | O_TRUNC, 0600);
+		if (pd->pd_fd < 0)
+			goto error_exit;
+	}
+
+	if (!pcapng_write_shb(pd, comment, hardware, os, user_application))
+		goto error_exit;
+
+	return pd;
+
+error_exit:
+	if (pd) {
+		if (pd->pd_fd < 0 && pd->pd_fd != STDOUT_FILENO)
+			close(pd->pd_fd);
+
+		free(pd);
+	}
+	return NULL;
+}
+
+/*****************************************************************************
+ * xpcapng_dump_close()
+ *****************************************************************************/
+void xpcapng_dump_close(struct xpcapng_dumper *pd)
+{
+	if (pd == NULL)
+		return;
+
+	if (pd->pd_fd < 0 && pd->pd_fd != STDOUT_FILENO)
+		close(pd->pd_fd);
+
+	free(pd);
+}
+
+/*****************************************************************************
+ * xpcapng_dump_flush()
+ *****************************************************************************/
+int xpcapng_dump_flush(struct xpcapng_dumper *pd)
+{
+	if (pd != NULL)
+		return fsync(pd->pd_fd);
+
+	errno = EINVAL;
+	return -1;
+}
+
+/*****************************************************************************
+ * pcapng_dump_add_interface()
+ *****************************************************************************/
+int xpcapng_dump_add_interface(struct xpcapng_dumper *pd, uint16_t snap_len,
+			       const char *name, const char *description,
+			       const uint8_t *mac, uint64_t speed,
+			       uint8_t ts_resolution, const char *hardware)
+{
+	if (!pcapng_write_idb(pd, name, snap_len, description, mac, speed,
+			      ts_resolution, hardware))
+		return -1;
+
+	return pd->pd_interfaces++;
+}
+
+/*****************************************************************************
+ * xpcapng_dump_enhanced_pkt()
+ *****************************************************************************/
+bool xpcapng_dump_enhanced_pkt(struct xpcapng_dumper *pd, uint32_t ifid,
+			       enum pcapng_epb_flags flags, uint64_t timestamp,
+			       uint32_t len, uint32_t caplen,
+			       const uint8_t *pkt, uint64_t dropcount,
+			       const char *comment, uint64_t *packetid,
+			       uint32_t *queue, int64_t *xdp_verdict)
```

Comment:
12 arguments? Wow. Could we maybe put these into a struct or something? The call to this becomes quite unreadable...

Reply:
Yes, I guess it grew on me :) I'll re-order it, and add a structure for the non mendatory options.